### PR TITLE
Airport Challenge Submission

### PR DESCRIPTION
**Pull request by Gwasanaethau**
My rather frustrated attempt at the Airport Challenge. Sorry it is not in any particularly great condition.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,62 +1,11 @@
...
 
-Instructions
----------
+The purpose of the Airport Challenge is to test the knowledge of and skills in BDD, TDD and OOP gained over the previous few weeks. The ideas is to build a working airport with aeroplanes based on the use-cases outlined in the [Instructions](./Instructions.md) over the weekend, and then submit the progress by 09:00 on Monday 23rd March.
 
-* Challenge time: rest of the day and weekend, until Monday 9am
-* Feel free to use google, your notes, books, etc but work on your own
-* If you have a partial solution, still check in a partial solution
-* You must submit a pull request to this repo with your code by 9am Monday morning
+I have really struggled with this challenge this weekend. I still cannot figure out how to translate the user stories into feature tests. I am also confused as to how feature tests and unit tests differ: in all examples I have read (including the Boris Bikes exercises) it seems the feature tests and unit tests test for exactly the same behaviour. I even struggled to translate the examples in the Boris Bikes exercises into something that I could use in the Airport Challenge.
```
Comment (sjmog) [inline README.md]: Is this a little clearer after the demonstrations today? How?

File: README.md [hunk 2]
```diff
@@ -1,62 +1,11 @@
...
 
-Steps
--------
+In addition, I have really struggled to find the correct usage for applying stubs to method returns to consistently mask out the required behaviour from a given method. Many examples give the required syntax, but none explain what each item refers to.
```
Comment (sjmog) [inline README.md]: I'll leave comments in-line for these then :)

File: README.md [hunk 3]
```diff
@@ -1,62 +1,11 @@
...
 
-1. Fill out your learning plan self review for the week: https://github.com/makersacademy/learning_plan (start by forking this repo, then edit week 1 - you can edit directly on Github)
-2. Fork this repo, and clone to your local machine
-3. Complete the following task:
+It may not seem like I have much work to show here, especially given the commit history, but rest assured that I devoted far too much time researching and asking questions, for most of which the results were not appropriate to add to any commit.
```
Comment (sjmog) [inline README.md]: That's the developer skill! 10x worse would be a mess of spaghetti code.

File: README.md [hunk 4]
```diff
@@ -1,62 +1,11 @@
...
-* Tests pass
-* Tests coverage is good
-* The code is elegant: every class has a clear responsibility, methods are short etc.
-
-Note that is a practice 'Tech Test' of the kinds that employers use to screen developer applicants.  More detailed submission requirements/guidelines are in [CONTRIBUTING.md](CONTRIBUTING.md)
-
-Finally, don’t overcomplicate things. This task isn’t as hard as it may seem at first. 
-
-* Finally submit a pull request before Monday at 9am with your solution or partial solution.  However much or little amount of code you wrote please please please submit a pull request before Monday at 9am
+I can only apologise for the current state of this project, but it is the best I have been able to do for the moment…
```
Comment (sjmog) [inline README.md]: Chill your beans :dancers:

File: lib/plane.rb [hunk 5]
```diff
@@ -1,2 +1,14 @@
+require 'weather'
+
 class Plane
+  include Weather
```
Comment (sjmog) [inline lib/plane.rb]: Awesome use of Weather as a module

File: lib/plane.rb [hunk 6]
```diff
@@ -1,2 +1,14 @@
...
+
+  def land
+    fail 'Weather Stormy' if stormy?
```
Comment (sjmog) [inline lib/plane.rb]: Love the terseness here

File: lib/plane.rb [hunk 7]
```diff
@@ -1,2 +1,14 @@
...
+    true
+  end
+
+  def take_off
+    true
```
Comment (sjmog) [inline lib/plane.rb]: I'd be expecting this to modify the *state* of the plane in some way (even more explicitly, if you put a `!` (bang) after it).

File: lib/weather.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+module Weather
+  def stormy?
+    Random.new.rand(100) < 5
```
Comment (sjmog) [inline lib/weather.rb]: Is it necessary to invoke the full `Random` library?

File: lib/weather.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+module Weather
```
Comment (sjmog) [inline lib/weather.rb]: Love the use of module here

File: spec/feature/atc_spec.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+require 'capybara/rspec'
+require 'plane'
+
+feature 'plane should not land at airport in certain circumstances –' do
+  let(:plane) { Plane.new }
```
Comment (sjmog) [inline spec/feature/atc_spec.rb]: Great use of `let` to store the object for later reference

File: spec/feature/atc_spec.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
+require 'capybara/rspec'
+require 'plane'
+
+feature 'plane should not land at airport in certain circumstances –' do
```
Comment (sjmog) [inline spec/feature/atc_spec.rb]: Really love this feature-naming convention. Can I steal it?

File: spec/feature/atc_spec.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  let(:plane) { Plane.new }
+
+  # As an air traffic controller
+  # So that I can avoid collisions
+  # I want to be able to prevent airplanes landing when the airport if full
+
+  scenario 'when airport is full' do
```
Comment (sjmog) [inline spec/feature/atc_spec.rb]: I think your main difficulty here is you're not integrating with an `Airport` class. Sometimes feature tests are called *Integration Tests* for this very reason - they detail the behaviour of interacting classes.

File: spec/feature/atc_spec.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    expect { plane.land }.to raise_error 'Airport Full'
+  end
+
+  # As an air traffic controller
+  # So that I can avoid accidents
+  # I want to be able to prevent airplanes landing when the weather is stormy
+
+  scenario 'when weather is stormy' do
+    allow(plane.land).to receive(:stormy?) { true }
```
Comment (sjmog) [inline spec/feature/atc_spec.rb]: This seems odd - why would a method receive another method, unless it was metaprogrammed?

File: spec/feature/atc_spec.rb [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    expect { plane.land }.to raise_error 'Weather Stormy'
```
Comment (sjmog) [inline spec/feature/atc_spec.rb]: Great use of `raise_error` :+1:

File: spec/feature/pilot_spec.rb [hunk 15]
```diff
@@ -0,0 +1,22 @@
+require 'capybara/rspec'
+require 'plane'
+
+# As a pilot
+# So that I can arrive at my specified destination
+# I would like to land my plane at the appropriate airport
+
+feature 'plane should land at airport' do
+  scenario 'plane should land at airport when told' do
+    expect(Plane.new.land).to be
```
Comment (sjmog) [inline spec/feature/pilot_spec.rb]: Try to think 'how would I do this in IRB?'. So for example, I'd say we should have

`plane = Plane.new`
`airport = Airport.new`
`plane.land(airport)`
`expect(plane).to be_landed # if plane has a predicate landed? method`

File: spec/feature/pilot_spec.rb [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+  end
+end
+
+# As a pilot
+# So that I can set off for my specified destination
+# I would like to be able to take off from the appropriate airport
+
+feature 'plane should take-off from airport' do
+  scenario 'plane should take-off from airport when told' do
+    expect(Plane.new.take_off).to be
```
Comment (sjmog) [inline spec/feature/pilot_spec.rb]: See above here. Think 'What would I need to do this in IRB?'

File: spec/plane_spec.rb [hunk 17]
```diff
@@ -1,25 +1,17 @@
...
-
-  xit 'can take off'
-
-  xit 'changes its status to flying after taking off'
-
+  let(:plane) { Plane.new }
+
+  it { expect(plane).to respond_to :land }
+  it 'should land if weather is not stormy' do
+    allow(plane.land).to receive(:stormy?) { false }
```
Comment (sjmog) [inline spec/plane_spec.rb]: Still confused as to why a method (`:land`) is receiving another method (`:stormy?`)

File: spec/plane_spec.rb [hunk 18]
```diff
@@ -1,25 +1,17 @@
...
+    expect(plane.land).to be
```
Comment (sjmog) [inline spec/plane_spec.rb]: You need some way for the plane to:

1) Land
2) Tell you its status (maybe something like `landed?`)

**Comment by sjmog** [comment]
There's a lot to love here, Mark. I think you're being very hard on yourself: you've mastered RSpec syntax, and have a good idea of how to construct Unit Tests. I think you're struggling to understand the role of Integration ('Feature') Tests in BDD. My main advice is to imagine your feature test as your automated interface. 

If you were making a cooking robot, you'd program a test to have it make 100 pancakes, and serve each one to a single tester so each pancake is verified to `be_tasty`. In this case, the interface is eating. The tester interacts with the pancake by eating it. In our airport challenge, our 'customer' will interact with the program through IRB. So that seems like a good place to start the design - "to fulfil the customer requirement, what would I have them type here?". You'll find natural objects drop out everywhere, and make for a great start to a class structure.

That said, awesome job with the reading, the researching, and (possibly) tying yourself in knots. That's what you're here to do! :+1:

**Comment by sjmog** [comment]
Don't forget to refactor this so it passes!

:construction:

File: spec/feature/atc_spec.rb (hunk 11 above)
Comment (Gwasanaethau) [inline spec/feature/atc_spec.rb]: Sure!

File: spec/feature/atc_spec.rb (hunk 13 above)
Comment (Gwasanaethau) [inline spec/feature/atc_spec.rb]: The ‘land’ method calls ‘stormy?’ to check if the weather is stormy or not. As ‘stormy?’ returns a random value, do I not need to stub that value out here? I think I have misunderstood how to use this.

**Pull request closed without merging**